packages/journey-manager/src/index.ts

```diff
@@ -371,25 +404,24 @@ export const run = async (props: RunProps): Promise<RunOutput> => {
     [],
   );
 
+  const appearSteps: StepWithQuery[] = Object.entries(triggers).reduce(
+    (prev: StepWithQuery[], [stepId, trigger]: [StepId, Trigger]) => {
+      if (trigger.event === "appear" && trigger.query != null) {
+        const newEntry: StepWithQuery = {
+          stepId,
+          query: decode(trigger.query),
+          urlCondition: trigger.urlCondition,
+          once: trigger.once,
+        };
+        return [...prev, newEntry];
+      }
+      return prev;
+    },
+    [],
+  );
```

Comment:
I'd suggest separating the generic filterMap out, it makes the logic easier to follow.

```suggestion
  function filterMap<A, B>(fn: A -> B | undefined | null, arr: A[]) : B[] {
     return arr.reduce((out, val) => {
        const res = fn(val);
        if (res != null) {
           out.push(res);
        }
        return res;
     }, []);
  }
  const appearSteps: StepWithQuery[] = filterMap(([stepId, trigger]: [StepId, Trigger]) => {
    if (trigger.event === "appear" && trigger.query != null) {
        const newEntry: StepWithQuery = {
          stepId,
          query: decode(trigger.query),
          urlCondition: trigger.urlCondition,
          once: trigger.once,
        }
    }, Object.entries(triggers));
```

Reply:
You already implemented one, I rewrote everything to use it.